conan/locks/backend_sqlite3.py

```diff
@@ -0,0 +1,69 @@
+import os
+from io import StringIO
+
+from conan.locks.backend import LockBackend
+from conan.locks.exceptions import AlreadyLockedException
+from conan.utils.sqlite3 import Sqlite3MemoryMixin, Sqlite3FilesystemMixin
+
+
+class LockBackendSqlite3(LockBackend):
+    # Sqlite3 backend to store locks. It will store the PID of every writer or reader before
+    # the can proceed to the resource (exclusive writer strategy).
+
+    LockId = int
+    _table_name = 'conan_locks'
+    _column_resource = 'resource'
+    _column_pid = 'pid'
+    _column_writer = 'writer'
+
+    def dump(self, output: StringIO):
+        with self.connect() as conn:
+            r = conn.execute(f'SELECT * FROM {self._table_name}')
+            for it in r.fetchall():
+                output.write(it)
+
+    def create_table(self, if_not_exists: bool = True):
+        guard = 'IF NOT EXISTS' if if_not_exists else ''
+        query = f"""
+        CREATE TABLE {guard} {self._table_name} (
+            {self._column_resource} text NOT NULL,
+            {self._column_pid} integer NOT NULL,
+            {self._column_writer} BOOLEAN NOT NULL CHECK ({self._column_writer} IN (0,1))
+        );
+        """
+        with self.connect() as conn:
+            conn.execute(query)
+
+    def try_acquire(self, resource: str, blocking: bool) -> LockId:
```

Comment:
According to docs, it's atomic, the `with self.connect() as conn:` should take care of it.

Reply:
I am not sure about this. The following code works, adding 2 lines to the table, both with a 1=blocking state over the same resource:

```python
 with connect() as conn1:
        with connect() as conn2:
            # Check if any is using the resource
            print("SELECT from conn1")
            result = conn1.execute(f'SELECT Pid, Writer '
                                    f'FROM MyTable '
                                    f'WHERE Resource = ?;', (resource,))

            print("SELECT from conn2")
            result2 = conn2.execute(f'SELECT Pid, Writer '
                                    f'FROM MyTable '
                                    f'WHERE Resource = ?;', (resource,))

            if result.fetchone():
                raise Exception("Resource {} locked".format(resource))

            if result2.fetchone():
                raise Exception("Resource {} locked".format(resource))

            # Check if a writer (exclusive) is blocking
            blocked = any([it[1] for it in result.fetchall()])
            if blocked:
                raise Exception("Resource {} locked".format(resource))

             # Check if a writer (exclusive) is blocking
            blocked = any([it[1] for it in result2.fetchall()])
            if blocked:
                raise Exception("Resource {} locked".format(resource))

            # Add me as a blocker, reader or writer
            blocking_value = 1 if blocking else 0

            result2 = conn2.execute(f'INSERT INTO MyTable '
                                    f'VALUES (?, ?, ?)', (resource, id(conn2), blocking_value,))

        result = conn1.execute(f'INSERT INTO MyTable '
                               f'VALUES (?, ?, ?)', (resource, id(conn1), blocking_value,))
        print("RESULT", result.lastrowid)
        print("RESULT2", result2.lastrowid)
```

Reply:
Yes, but this is not what is written in the PR...

Reply:
Ok, it doesn't block. The transaction is rolled back in case of an exception. It is in the TODO list.

Reply:
I keep having a loot to it, but looks like, something like this will do the trick:

```python
    @contextmanager
    def connect(self):
        conn = sqlite3.connect(....)
        conn.isolation_level = None
        try:
            conn.execute('begin')
            yield conn.cursor()
            conn.execute("commit")
        except Exception as e:
            conn.execute("rollback")
            raise e
        finally:
            conn.close()
```